src/uri.rs

```diff
@@ -186,6 +182,33 @@ impl Rsync {
         &self.as_str()[self.module_start..(self.path_start - 1)]
     }
 
+    /// Returns the URI’s module.
+    ///
+    /// The module is identical to the a URI with an empty path.
+    pub fn module(&self) -> &str {
+        &self.as_str()[..self.path_start]
+    }
+
+    /// Returns the URI’s canonical module.
+    ///
+    /// This is the same as the module but with the authority in canonical
+    /// form.
+    pub fn canonical_module(&self) -> Cow<str> {
```

Comment:
Unless I'm misreading, in `from_bytes()` you permit any case for `RsYnC://` yet here you explicitly use `rsync://` in lowercase (and then apply make_ascii_lowercase() to it lowering it if it were not already lowered), so in effect you also lowercase the scheme component too.

There is also a very small chance that having two separate occurences of the `rsync://` static string in the code could lead to mistakes where one is changed and the other not, perhaps using a single constant in both places would be a minor but worthwhile improvement?

Also, did you consider iterating over the value returned by `module()` and map uppercase to lowercase on that value rather than construct a new string entirely (potentially not being exactly the same as the type instance currently represents) as you do here?

Reply:
The scheme identifier of a URI is case-insensitive (i.e, `RsYnC` is allowed and equivalent to `rsync`), so in order to get a canonical URI, I need to lowercase the scheme, too. (In fact RFC 3986 defines the canonical form of the scheme to be lowercase).

I think using the string literal here is fine – why would you want to ever change it?

As for changing the contained value itself, `Bytes` is a potentially shared type (kind of like an `Arc<Box<[u8]>>` on steroids) and therefore immutable. So I have to clone the data in order to lowercase it.